waspc/src/Analyzer/Lib.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Analyzer.Lib
+  ( Lib (..),
+    empty,
+    addEnumType,
+    addDeclType,
+    EnumType (..),
+    DeclType (..),
+    IsEnum,
+    IsDecl,
+  )
+where
+
+import Analyzer.Type (Type)
+import qualified Data.HashMap.Strict as M
+
+newtype EnumType = EnumType {variants :: [String]}
+
+newtype DeclType = DeclType {declType :: Type}
+
+-- | Contains information for type-checking declarations and enums and for
+--   constructing instances of Haskell types corresponding to the declarations
+--   and enums.
+data Lib = Lib
+  { decls :: M.HashMap String DeclType,
+    enums :: M.HashMap String EnumType
+  }
+
+-- | A library with no types
+empty :: Lib
+empty = Lib {decls = M.empty, enums = M.empty}
+
+-- | Add a declaration type to a library. Requires the type to be in the form
+--   of a Wasp decl. See "IsDecl" for requirements.
+--
+--  Example:
+--
+--  @
+--  {-# LANGUAGE TypeApplications, DeriveGeneric #-}
+--  data App = App { title :: String } deriving Generic
+--  let exampleLib = addDeclType @App empty
+--  @
+addDeclType :: forall typ. (IsDecl typ) => Lib -> Lib
+addDeclType lib =
+  let decl = DeclType {declType = _declType @typ}
+   in lib {decls = M.insert (_declName @typ) decl $ decls lib}
+
+-- | Add an enum type to a library. Requires the type to be in the form
+--   of a Wasp enum. See "IsEnum" for requirements.
+--
+--   Example:
+--
+--   @
+--   {-# LANGUAGE TypeApplications, DeriveGeneric #-}
+--   data ParamType = StringParam | NumberParam deriving Generic
+--   let exampleLib = addEnumType @ParamType empty
+--   @
+addEnumType :: forall typ. (IsEnum typ) => Lib -> Lib
+addEnumType lib =
+  let enum = EnumType {variants = _enumVariants @typ}
+   in lib {enums = M.insert (_enumName @typ) enum $ enums lib}
+
+-- | Encodes the requirements of a Wasp decl.
+--
+--   For an instance to be created for a type with a `Generic` instance,
+--   - The type must be an ADT with one constructor
+--   - The type must have one field OR use record syntax
+--
+--   Some assumptions are required of `_declType` and `a`:
+--   - If `_declType` is a `Dict`, then
```

Comment:
Maybe my understanding of "if and only if" is not great, but I thought that `if and only if A is true, then B is true` means that B can't be true unless A is true, and if A is true, then you can be sure B is true -> so it goes both direction, B being true means A is true, and vice versa. So A and B are both true or both false. I looked up Wikipedia now and from what I got this is how "if and only if" ("iff") is defined.

And here I thought: if _declType is Dict, then `a` must use record syntax. On the other hand, if `a` uses record syntax, _declType must be Dict (in order for it all to make sense). So that is "if and only if", right? As for other two conditions, hm I have to admit I am not sure how to think about them but I thought they fall under the same logic.

Sorry if I got smth wrong here, if pls tell me.

Reply:
I get what you mean now. I've changed it to "if and only if" here.